### PR TITLE
Re-cost columnar table index scan paths

### PR DESCRIPTION
**Pull request by onurctirtir**
(Rebased onto #5058)

With the changes in this pr, we adjust the cost estimate done by indexAM for `IndexPath` according to columnar tables when the index is on a columnar table.
This is because, the way indexAM estimates the cost is not appropriate for indexes on columnar tables.
The most basic reason is that indexAM assumes we will only need to read single page to access a single tuple of the table.
On the other hand for columnar tables, we read the whole stripe from disk for a single tuple too, regardless of the optimization done in #5058.

Note that we don't simply assign startup / total costs but we add the cost estmated by us to the cost estimated by indexAM.
This is because we need to take "the cost due to index data-structure traversal" into account too.

Before explaining the logic that we follow for `IndexPath`, let's first summarize what we were / are doing for `ColumnarCustomScan`:
```math
X <- cost for reading single column of single stripe // 1
cost = X * (number of columns after projection pushdown) // 2
cost = cost * (number of stripes that relation has) // 3
```

The logic that we follow to calculate the additional cost for index scan is as follows:
```math
X <- cost for reading single column of single stripe // same as 1 above
cost = X * (number of columns that relation has) // index scan cannot do projection pushdown, so different than 2 above
cost = cost * (estimated number of stripes that we need to read)
```

where, we calculate `estimated number of stripes that we need to read` as follows:

```math
indexCorrelation, indexSelectivity <- calculate by using amcostestimate_function
estimatedReadRows = (relation row count) * indexSelectivity

minEstimateStripeReads = estimatedReadRows / (average stripe row count) // full correlation, we will not do any redundant stripe reads
maxEstimateStripeReads = estimatedReadRows // no correlation, we will read a different stripe for each tuple

complementCorrelation = 1 - abs(indexCorrelation)
estimatedStripeCount = minEstimateStripeReads +
                       complementCorrelation  * (maxEstimateStripeReads  - minEstimateStripeReads)
```

------

Note that this pr doesn't re-cost sequential scan paths, which we should do in another pr for the cases where `columnar.enable_custom_scan` is set to `'off'`.

------

TODO:
- [x] Add tests to show when we choose index scan / columnar custom scan

### CODE REVIEW

File: src/backend/columnar/columnar_customscan.c
Comment (jeff-davis) [inline src/backend/columnar/columnar_customscan.c:371]: Should we be counting the number of stripe reads, or chunk reads, or both? I'm inclined to think chunk reads are more important because of the decompression, but stripe reads do have some overhead as well due to missing blocks in the cache and not benefiting from readahead.

File: src/backend/columnar/columnar_customscan.c
Comment (jeff-davis) [inline src/backend/columnar/columnar_customscan.c:380]: I don't think "inverse" is the right word here.

File: src/backend/columnar/columnar_customscan.c
Comment (jeff-davis) [inline src/backend/columnar/columnar_customscan.c:383]: `maxStripeReadCount - minStripeReadCount` ≈ `maxStripeReadCount` in most cases. It might be worth pointing that out.

I like how you are trying to capture the actual costs, but in practice, I think we need the absIndexCorrelation to be *really* close to one, or the table to be just a couple chunks, to ever choose index scan. I think it's worth calling out in a comment that we either have a correlation > 0.999, or we might as well forget it.

File: src/backend/columnar/columnar_customscan.c
Comment (onurctirtir) [reply src/backend/columnar/columnar_customscan.c:380]: you're right, maybe complementCorrelation

File: src/backend/columnar/columnar_customscan.c
Comment (onurctirtir) [reply src/backend/columnar/columnar_customscan.c:371]: Initially I was thinking of calculating the cost based on number of chunk groups being decompressed.

However, reading the whole stripe (regardless of how many chunk groups we need to decompress) for a single tuple seemed to be the heavy weight part to me.

Also realized that, for custom scan, we were only taking number of blocks read from disk, not the decompressed chunk groups.

For those reasons, I followed a similar approach for index scan too.

On the other hand, ideally, we should also consider chunk group decompression cost for both custom scan & index scan.
But I think we can ignore that for the time being.

What do you think @jeff-davis ?

File: src/backend/columnar/columnar_customscan.c
Comment (jeff-davis) [inline src/backend/columnar/columnar_customscan.c:382]: Can you word more clearly? What is "mostly not the case in reality"?

There are a lot of cases where we'd expect index correlation to be 1 (or extremely close), like if someone creates an index on the timestamp. They may not really need the index in that case, because chunk group filtering is going to also be very effective. But it is a case where the correlation is 1.

I'd say something like: "In practice, we only want to do an index scan when the correlation is 1 (or extremely close to it), or when the absolute number of tuples returned is very small. Other cases will have a prohibitive cost."

File: src/backend/columnar/columnar_customscan.c
Comment (jeff-davis) [inline src/backend/columnar/columnar_customscan.c:257]: Should we include a comment about bitmap index scans here?

**Review by jeff-davis — APPROVED** [review]

File: src/backend/columnar/columnar_customscan.c
Comment (onurctirtir) [reply src/backend/columnar/columnar_customscan.c]: done

File: src/backend/columnar/columnar_customscan.c
Comment (onurctirtir) [reply src/backend/columnar/columnar_customscan.c:263]: done

**Comment by onurctirtir** [comment]
Thank you @jeff-davis for all the comments.
I will not merge this pr until merging #5058, since this pr is rebased onto that.

**Pull request merged**
